### PR TITLE
feat(Instagram): add privacy mode, more details

### LINKED ISSUE #5529: Privacy mode for Instagram
### Presence name

Instagram

### Description

Privacy mode for instagram, like facebook

### Alternatives

Nope

### Additional context

Nope

### PR DESCRIPTION
**Pull request by 7loud**
**Describe the changes in this pull request:**
- add Privacy Mode (Resolves #5529)
- add more details

<details>
<summary> Proof showing the creation/modification is working as expected </summary>
<br>

<hr>

*Button label was corrected*
<hr>

</details>

### CODE REVIEW

File: websites/I/Instagram/presence.ts [hunk 1]
```diff
@@ -1,37 +1,176 @@
...
-		presenceData.details = "Viewing a story";
-		presenceData.state = pathName;
-	} else if (document.location.pathname.startsWith("/accounts")) {
+			largeImageKey: "logo"
+		},
+		setting = {
+			privacy: await presence.getSetting<boolean>("privacy"),
+			elapsedTime: await presence.getSetting<boolean>("elapsedTime"),
+			postImage: await presence.getSetting<boolean>("postImage")
+		},
```
Comment (Bas950) [inline websites/I/Instagram/presence.ts:14]: promise.all?

**Review by Bas950 — DISMISSED** [review]

**Review by EncryptedDev — DISMISSED** [review]

File: websites/I/Instagram/presence.ts [hunk 2]
```diff
@@ -1,37 +1,190 @@
...
+			? "Viewing a Story"
+			: `Viewing ${path[2]}'s Story`;
+
+		if (time && time.getAttribute("datetime")) {
+			presenceData.state = getDateString(
+				new Date(time.getAttribute("datetime"))
+			);
+		}
+
+		if (!setting.privacy) {
```
Comment (ImRodry) [inline websites/I/Instagram/presence.ts:32]: spare us some checks and run this at the end (aka delete the unwanted props if true)

File: websites/I/Instagram/presence.ts [hunk 3]
```diff
@@ -1,37 +1,190 @@
...
-		const [, pathName] = document.location.pathname.split("/");
-		presenceData.details = "Viewing a profile";
-		presenceData.state = pathName;
+	else if (profileName?.textContent === path[1]) {
+		const profilePicture =
+			document.querySelector<HTMLImageElement>("img._6q-tv");
+
+		presenceData.details = setting.privacy
+			? "Viewing a Profile"
+			: "Viewing a Profile:";
```
Comment (ImRodry) [inline websites/I/Instagram/presence.ts:96]: please don't duplicate this much text, use template strings instead

**Review by ImRodry — CHANGES_REQUESTED** [review]

File: websites/I/Instagram/presence.ts [hunk 4]
```diff
@@ -1,37 +1,190 @@
...
+				"div.QY4Ed.P0xOK input.focus-visible"
+			)?.value;
+		}
+	}
+
 	presence.setActivity(presenceData);
 });
+
+function getDateString(date: Date) {
+	const seconds = Math.abs(Date.now() - +date) / 1000;
```
[the remark this answers is not in the document]
Comment (N0chteil) [reply websites/I/Instagram/presence.ts:129]: image

[the remark this answers is not in the document]
Comment (N0chteil) [reply websites/I/Instagram/presence.ts:129]: image

[the remark this answers is not in the document]
Comment (ImRodry) [reply websites/I/Instagram/presence.ts:129]: in that case
```suggestion
	const seconds = Math.abs(Date.now() - date.getTime()) / 1000;
```

File: websites/I/Instagram/presence.ts [hunk 5]
```diff
@@ -1,37 +1,190 @@
...
+
+	if (setting.elapsedTime) presenceData.startTimestamp = timestamp;
+
+	if (!path[1]) presenceData.details = "Viewing the Homepage";
+	else if (pathname.startsWith("/stories")) {
+		const time = document.querySelector("time.BPyeS.Nzb55"),
+			video = document.querySelector("video");
+
+		presenceData.details = setting.privacy
+			? "Viewing a Story"
```
[the remark this answers is not in the document]
Comment (N0chteil) [reply websites/I/Instagram/presence.ts:23]: It's uppercased everywhere on Instagram

File: websites/I/Instagram/presence.ts [hunk 6]
```diff
@@ -1,37 +1,190 @@
...
+
+async function getSettings() {
+	const settings = await Promise.all([
+			await presence.getSetting<boolean>("privacy"),
+			await presence.getSetting<boolean>("elapsedTime"),
+			await presence.getSetting<boolean>("postImage")
+		]),
+		names = ["privacy", "elapsedTime", "postImage"],
+		obj: {
+			[key: string]: boolean;
```
Comment (ImRodry) [inline websites/I/Instagram/presence.ts:182]:
```suggestion
			[key in typeof names[number]]?: boolean;
```

File: websites/I/Instagram/presence.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ImRodry) [reply websites/I/Instagram/presence.ts:23]: Example?

File: websites/I/Instagram/presence.ts [hunk 7]
```diff
@@ -1,37 +1,199 @@
...
+			presenceData.buttons = [
+				{
+					label: "View Story",
+					url: `https://www.instagram.com/stories/${path[2]}/${path[3]}`
+				}
+			];
+		} else {
+			delete presenceData.startTimestamp;
+			delete presenceData.endTimestamp;
+			delete presenceData.buttons;
```
Comment (ImRodry) [inline websites/I/Instagram/presence.ts:49]: I meant this on the entire presence, this else statement is useless here

File: websites/I/Instagram/presence.ts [hunk 8]
```diff
@@ -1,37 +1,199 @@
...
+			return `${Math.floor(seconds / 604800)} week${
+				seconds === 604800 ? "" : "s"
+			} ago`;
+		case seconds < 29030400:
+			return `${Math.floor(seconds / 2419200)} month${
+				seconds === 2419200 ? "" : "s"
+			} ago`;
+		default:
+			return `${Math.floor(seconds / 29030400)} year${
+				seconds === 29030400 ? "" : "s"
```
Comment (ImRodry) [inline websites/I/Instagram/presence.ts:162]: I’m sure this isn’t right, it’s very unlikely for the variable to have this specific value

File: websites/I/Instagram/presence.ts [hunk 9]
```diff
@@ -1,37 +1,190 @@
...
-	if (document.location.pathname === "/")
-		presenceData.details = "Viewing the homepage";
-	else if (document.location.pathname.startsWith("/stories")) {
-		const [, , pathName] = document.location.pathname.split("/");
-		presenceData.details = "Viewing a story";
-		presenceData.state = pathName;
-	} else if (document.location.pathname.startsWith("/accounts")) {
+			largeImageKey: "logo"
+		},
+		setting = await getSettings(),
```
[the remark this answers is not in the document]
Comment (ImRodry) [reply websites/I/Instagram/presence.ts]: And this was resolved because?

File: websites/I/Instagram/presence.ts [hunk 10]
```diff
@@ -1,37 +1,179 @@
...
+		obj: {
+			[key in typeof names[number]]?: boolean;
+		} = {};
+
+	names.forEach((name, i) => {
+		obj[name] = settings[i];
+	});
+
+	return obj;
+}
```
Comment (EncryptedDev) [inline websites/I/Instagram/presence.ts:179]: What's the point in this?

Comment (N0chteil) [reply websites/I/Instagram/presence.ts:179]: https://github.com/PreMiD/Presences/pull/5641#discussion_r790271040

**Review by EncryptedDev — DISMISSED** [review]

File: websites/I/Instagram/presence.ts (hunk 10 above)
Comment (Bas950) [reply websites/I/Instagram/presence.ts:179]: you dont need this whole function tho, just deconstruct the array instead of putting it in an object

**Review by Bas950 — APPROVED** [review]

**Review by EncryptedDev — APPROVED** [review]

**Pull request merged**